### ISSUE TITLE
Allow multiple impls for the same struct in different files

### ISSUE THREAD

**Issue by daniellga** [extendr/extendr#538]
Would it be possible for you to implement multiple `impl` to the same struct in different files? Here are the errors I get.

lib.rs
```
use extendr_api::prelude::*;

mod test1;
mod test2;

extendr_module! {
    mod library_name;
    use test1;
    use test2;
}

```

test1.rs
```
use extendr_api::prelude::*;

pub struct Lol;

#[extendr]
impl Lol {
    fn ttt() {}
}

extendr_module! {
    mod test1;
    impl Lol;
}

```

test2.rs
```
use extendr_api::prelude::*;

use crate::test1::Lol;

#[extendr]
impl Lol {
    fn zzz() {}
}

extendr_module! {
    mod test2;
    impl Lol;
}
```

Error:
```
error[E0119]: conflicting implementations of trait `extendr_api::FromRobj<'_>` for type `&Lol`
 --> src/test2.rs:5:1
  |
5 | #[extendr]
  | ^^^^^^^^^^ conflicting implementation for `&Lol`
  |
 ::: src/test1.rs:5:1
  |
5 | #[extendr]
  | ---------- first implementation here
  |
  = note: this error originates in the attribute macro `extendr` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0119]: conflicting implementations of trait `extendr_api::FromRobj<'_>` for type `&mut Lol`
 --> src/test2.rs:5:1
  |
5 | #[extendr]
  | ^^^^^^^^^^ conflicting implementation for `&mut Lol`
  |
 ::: src/test1.rs:5:1
  |
5 | #[extendr]
  | ---------- first implementation here
  |
  = note: this error originates in the attribute macro `extendr` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0119]: conflicting implementations of trait `From<Lol>` for type `extendr_api::Robj`
 --> src/test2.rs:5:1
  |
5 | #[extendr]
  | ^^^^^^^^^^ conflicting implementation for `extendr_api::Robj`
  |
 ::: src/test1.rs:5:1
  |
5 | #[extendr]
  | ---------- first implementation here
  |
  = note: this error originates in the attribute macro `extendr` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0119]: conflicting implementations of trait `From<&Lol>` for type `extendr_api::Robj`
 --> src/test2.rs:5:1
  |
5 | #[extendr]
  | ^^^^^^^^^^ conflicting implementation for `extendr_api::Robj`
  |
 ::: src/test1.rs:5:1
  |
5 | #[extendr]
  | ---------- first implementation here
  |
  = note: this error originates in the attribute macro `extendr` (in Nightly builds, run with -Z macro-backtrace for more info)
```

**Comment by multimeric** [comment]
Here's the issue. When we run the `#[extendr]` on an `impl SomeStruct` block {}, we don't just generate the function wrappers, we also generate the "core" extendr translation for the struct itself: https://github.com/extendr/extendr/blob/fc677380fa03c638e7913fdd96140da9ff1a5ad4/extendr-macros/src/extendr_impl.rs#L108C3-L187. Thus, if we do so a second time in the same application, we have duplicate copies of these core implementations.

Considering this, my opinion is that it was a mistake to generate this code when we process an impl block. I think the ideal behaviour would be to `#[extendr]` *on the struct definition*, which generates this stuff *once only*, and then you can do as many `impl` blocks as you like, and it will only generate the function wrappers when you do.

However, if we made this change it would break backwards compatibility, so in the short term I propose these changes:
1. We add an `#[extendr]` option to `impl` blocks, that disables the "core" struct code. e.g. `#[extendr(only_funcs = true)]`
2. We allow putting `#[extendr]` over a struct, which *only* outputs the core struct code, similar to what happens when you put `#[extendr]` over an empty `impl` block now.

**Comment by eitsupi** [comment]
I know you already have multiple breaking changes merged into the default branch now.
Do you have any plans to address this issue?

As it seems to be very painful to have only one impl block in a project like r-polars where there is a large number of functions.

Of course, now there is savvy that allows multiple impl blocks, so it may just be that the user can choose savvy instead of extendr having to deal with this.

**Comment by eitsupi** [comment]
> I'd really like you to attempt to be a bit more cordial here. I reply to every ping from you, and I definitely want to make extendr friendly to you.

Sorry, I think it's my poor English, but I don't know what you want from me.

I simply wanted to show that this issue, which appears to have not progressed in over a year, may be important.

> Finally, I would really appreciate it, if you could leave reviews for some open PRs. I can only work on new things, when some of my open PRs get merged. You contributed before, and thus, your opinion is very important to us.

It is great to see that you have been actively updating extendr lately, but unfortunately I am not knowledgeable enough to look at your code and understand the content.

**Comment by CGMossa** [comment]
I've just looked at it again. The issue isn't the breaking change. The issue is that we generate the wrapper by `extendr_module`, and there, there is one line for an `impl`-block. That's what calls the wrapper-generating function for that particular block. I don't know how to register all blocks.

The blocks would have to be named or something for me to make that work. The generation of macros isn't really something to depend on.

I was thinking to make a `build.rs` file that analyses user packages and stores such information.. But that was too big for me to think about. It would be using `syn` to build a hashmap of type-information.

**Comment by CGMossa** [comment]
Also @eitsupi, as far as I understand it, savvy doesn't write the R wrapper for you. extendr does. So unless I can figure out a way to have a spot (in proc-macro) that can locate all impl-blocks, I really don't see what to do about this.

**Comment by eitsupi** [comment]
> Also @eitsupi, as far as I understand it, savvy doesn't write the R wrapper for you. extendr does. So unless I can figure out a way to have a spot (in proc-macro) that can locate all impl-blocks, I really don't see what to do about this.

Sorry, I do not know about the implementation details.
I don't see why something that seems to work for pyo3 and napi can't be done with extendr.

**Comment by CGMossa** [comment]
I had a look at `napi`, and it doesn't need to register individual functions like we do (as far as I can tell).

Python have to do some method registration.
Looking at pyo3, it actually requires you to write a bit more elaborate export code than `extendr`, see <https://pyo3.rs/v0.21.2/function>.

I'm not yet convinced, that others have solved this problem. I'm looking for an idea, and I've posted about this, and my ideas for it, on the discord _a lot_. 

Please let me know of more examples that you know of, that could help with this. I'll be looking out for an answer too.

**Comment by multimeric** [comment]
Had you considered my original suggestion?

**Comment by CGMossa** [comment]
> Had you considered my original suggestion?

Yes! I've considered it multiple times. If I move those impls to a macro that only works on `struct`, that still doesn't solve my problem: `extendr_module` sees an `impl`-block, it calls _one_ impl block wrapper-generating function, and there is no way for me to detect other blocks and their impl-block wrapper generating function.

Each impl-block generates a function that returns the meta-data for wrapper functions. It gets named _something_ that the `extendr_module!` relies on.

**Comment by CGMossa** [comment]
I've investigated this further, and I believe what we need to do, is introduce new syntax, amongst other things:

- Mandate `#[extendr]` on `struct` from now on. 

```rs
#[extendr]
struct Model;

#[extendr(label = "name1")]
impl Model {
// [...]
}

#[extendr(label = "name2")]
impl Model {
// [...]
}

extendr_module! {
  #[extendr(label = "name1")]
  impl Model;
  #[extendr(label = "name2")]
  impl Model;
}
```

I am very positive that this might work. Recently, I have experimented with adding `extern "C" fn`-support to extendr, so I've gone through the wrapper generation. I need to make a few improvements to the derive-macros, but this is very feasible in the near future.

**Comment by JosiahParry** [comment]
This is available in the development version of extendr.
